src/nvim/api/vim.c

```diff
@@ -207,18 +208,38 @@ String nvim_replace_termcodes(String str, Boolean from_part, Boolean do_lt,
   return cstr_as_string(ptr);
 }
 
-String nvim_command_output(String str, Error *err)
+/// Executes an ex-command and returns its (non-error) output.
+/// Shell |:!| output is not captured.
+///
+/// On parse error: forwards the Vim error; does not update v:errmsg.
+/// On runtime error: forwards the Vim error; does not update v:errmsg.
+///
+/// @param command  Ex-command string
+/// @param[out] err Error details (Vim error), if any
+String nvim_command_output(String command, Error *err)
   FUNC_API_SINCE(1)
 {
-  do_cmdline_cmd("redir => v:command_output");
-  nvim_command(str, err);
-  do_cmdline_cmd("redir END");
+  Array args = ARRAY_DICT_INIT;
+  ADD(args, STRING_OBJ(command));
+  ADD(args, STRING_OBJ(cstr_as_string("silent")));
 
+  Object rv = nvim_call_function(cstr_as_string("execute"), args, err);
+  api_free_array(args);
   if (ERROR_SET(err)) {
+    assert(rv.type == kObjectTypeNil);
     return (String)STRING_INIT;
   }
 
-  return cstr_to_string((char *)get_vim_var_str(VV_COMMAND_OUTPUT));
+  assert(rv.type == kObjectTypeString);
+  // execute() always(?) prepends a newline; remove it.
+  if (rv.data.string.size > 1) {
```

Comment:
This looks like cargo-cult magic. A common implementation will avoid this.

Reply:
It's "I don't care why, but it's guarded by an assert()".

Reply:
You can mark it with a TODO if noone cares to fix it right now. In the long run "the behaviour of the first error message inside a capture" doesn't sound like like something we should express uncertainty about, rather it should just do the intended thing directly.

Reply:
The behavior of execute () is vim-compatible so I don't think it can be changed. 

The `assert(...='\n')` below here serves to document the assumption.

Reply:
The behavior of `execute()` is already configurable, it wouldn`t be strange for a common implementation to also have this as a flag.

I might look into the relevant parts anyway when revisiting #7466 (which in some sense is `command_output` on steroids, but continuously). If we find a better way to treat newlines, it wouldn't be completely unrealistic for vim to also implement it? (either behind a new `silent` value, or change the default on the same theory of not bug-for-bug compatibility).